File: c++/47_Permutations_II.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int> > permuteUnique(vector<int>& nums) {
    	int n = nums.size();
    	sort(nums.begin(), nums.end());
		vector<vector<int> > res;
    	if (n <= 1) {
    		res.push_back(nums);
    		return res;
    	}
    	int last = nums[0] - 1; // make sure they are not the same
    	for (int i = 0; i < n; i++) {
    		if (last == nums[i]) // recursion only when new element appears
    			continue;
    		vector<int> tmp = nums;
    		tmp.erase(tmp.begin() + i);
    		vector<vector<int> > res_part = permuteUnique(tmp);
    		for (int j = 0; j < res_part.size(); j++) {
    			res_part[j].push_back(nums[i]);
    			res.push_back(res_part[j]);
    		}
    		last = nums[i];
    	}
    	return res;
    }
};

File: python/47_Permutations_II.py
#!/usr/bin/python
import sys
import os

class Solution(object):
    def permuteUnique(self, nums):
        n, res = len(nums), []
        if n <= 1:
        	return [nums]

        nums, last = sorted(nums), "."
        for i in range(n):
        	if last == nums[i]:
        		continue
        	for arr in self.permuteUnique(nums[:i] + nums[i + 1:]):
        		res.append(arr + [nums[i]])
        	last = nums[i]
        return res


        


def main():
    s = Solution()
    print s.permuteUnique([1, 1, 2])

if __name__ == "__main__":
    main()
